src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -1501,5 +1504,115 @@ private interface IWebSocketReceiveResultGetter<TResult>
             public WebSocketReceiveResult GetResult(int count, WebSocketMessageType messageType, bool endOfMessage, WebSocketCloseStatus? closeStatus, string closeDescription) =>
                 new WebSocketReceiveResult(count, messageType, endOfMessage, closeStatus, closeDescription);
         }
+
+        private class EnsureBuffer : IValueTaskSource
+        {
+            private readonly ManagedWebSocket _webSocket;
+            private readonly Action _onComplete;
+            private ManualResetValueTaskSourceCore<VoidResult> _mrvts;
+
+            private int _minimumRequiredBytes;
+            private CancellationToken _cancellationToken;
+            private bool _throwOnPrematureClosure;
+
+            private ValueTaskAwaiter<int> _awaiter;
+
+            public EnsureBuffer(ManagedWebSocket webSocket)
+            {
+                _webSocket = webSocket;
+                _onComplete = new Action(OnComplete);
+            }
+
+            public ValueTask EnsureBufferContainsAsync(int minimumRequiredBytes, CancellationToken cancellationToken, bool throwOnPrematureClosure)
+            {
+                _minimumRequiredBytes = minimumRequiredBytes;
+                _cancellationToken = cancellationToken;
+                _throwOnPrematureClosure = throwOnPrematureClosure;
+
+                EnsureBufferContains();
+
+                return new ValueTask(this, _mrvts.Version);
```

Comment:
Doesn't this mean that if the stream.ReadAsync completes synchronously (which should be relatively common) that we're now going to go through the more expensive paths?

Reply:
Added fast path